text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,426 @@
+
+Field types serve one important purpose, because they are types, they can implement traits, and thus be generalized over. This allows for powerful compile time guarantees that field types are being used correctly.
+
+### `trait Field` and `type FieldDescriptor`
+
+In order to generalize over field types, we have the `trait Field` 
+
+```rust
+trait Field {
+    /// The type of the type that the field comes from
+    type Parent: ?Sized;
+
+    /// The type of the field itself
+    type Type: ?Sized;
+
+    const FIELD_DESCRIPTOR: FieldDescriptor<Self>;
+}
+
+struct FieldDescriptor<F: Field + ?Sized> { .. }
+```
+
+`FieldDescriptor` is an opaque type that will store some metadata about how to convert from a `*const Field::Parent` to a `*const Field::Type`. There is no way to safely construct `FieldDescriptor` from user code on Stable Rust until Rust gets a defined stable type layout for `repr(Rust)` types.
+
+The `Field` trait will allow generalizing over field types, and thus allow other apis to be created, for example...
+
+### `*const T`/`*mut T` methods
+
+We will add the following methods to raw pointers
+
+```rust
+// details about why we need both and what they do exactly in Reference-level explanation
+
+impl<T: ?Sized> *const T {
+    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type;
+    
+    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *const F::Type;
+
+    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
+    
+    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *const F::Parent;
+}
+
+impl<T: ?Sized> *mut T {
+    pub unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
+    
+    pub fn wrapping_project<F: Field<Parent = T>>(self, field: F) -> *mut F::Type;
+    
+    pub unsafe fn inverse_project_unchecked<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
+    
+    pub fn inverse_wrapping_project<F: Field<Type = T>>(self, field: F) -> *mut F::Parent;
+}
+```
+
+These will allowing projections through raw pointers without dereferencing the raw pointer. This is useful for building projections through other abstractions like smart pointers (`Rc<T>`, `Pin<&T>`)!
+
+This is the extent of the core api of this RFC.
+
+Using this we can do something like this
+```rust
+struct Foo {
+    bar: Bar,
+    age: u32
+}
+
+struct Bar {
+    name: String,
+    id: i32
+}
+
+let x : Foo = ...;
+let y : *const Foo = &x;
+
+let y_bar_name: *const String = unsafe { y.project_unchecked(Foo.bar).project_unchecked(Bar.name) };
+```
+
+In the end `y_bar_name` will contain a pointer to `x.bar.name`, all without dereferencing a single pointer! (Given that this is a verbose, we may want some syntax for this, but that is out of scope for this RFC)
+
+But, we can build on this foundation and create a more power abstraction, to generalize this project notion to smart pointers.
```

Comment:
This seems to be out of place now, as the more powerful abstraction isn't provided by this RFC anymore.

Reply:
True, but I still think that it is worth pointing out. Also, I just moved `Project` to the future possibilities section so people can still so what a project trait could look like, and how it would work with this proposal.